File: backend/app/models/alarm.py
"""
告警模型
"""
from datetime import datetime
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Float, Text, ForeignKey, Date

from ..core.database import Base


class AlarmThreshold(Base):
    """告警阈值配置表"""
    __tablename__ = "alarm_thresholds"

    id = Column(Integer, primary_key=True, autoincrement=True)
    point_id = Column(Integer, ForeignKey("points.id"), nullable=False, comment="点位ID")
    threshold_type = Column(String(20), nullable=False, comment="阈值类型: high_high/high/low/low_low/equal/change")
    threshold_value = Column(Float, comment="阈值")
    alarm_level = Column(String(20), default="minor", comment="告警级别: critical/major/minor/info")
    alarm_message = Column(String(200), comment="告警消息")
    delay_seconds = Column(Integer, default=0, comment="延迟触发(秒)")
    dead_band = Column(Float, default=0, comment="死区(回差)")
    is_enabled = Column(Boolean, default=True, comment="是否启用")
    priority = Column(Integer, default=0, comment="优先级")
    created_at = Column(DateTime, default=datetime.now, comment="创建时间")
    updated_at = Column(DateTime, default=datetime.now, onupdate=datetime.now, comment="更新时间")


class Alarm(Base):
    """告警记录表"""
    __tablename__ = "alarms"

    id = Column(Integer, primary_key=True, autoincrement=True)
    alarm_no = Column(String(50), unique=True, nullable=False, comment="告警编号")
    point_id = Column(Integer, ForeignKey("points.id"), nullable=False, comment="点位ID")
    threshold_id = Column(Integer, ForeignKey("alarm_thresholds.id"), comment="阈值配置ID")
    alarm_level = Column(String(20), nullable=False, comment="告警级别")
    alarm_type = Column(String(20), comment="告警类型: threshold/communication/system")
    alarm_message = Column(Text, nullable=False, comment="告警消息")
    trigger_value = Column(Float, comment="触发值")
    threshold_value = Column(Float, comment="阈值")
    status = Column(String(20), default="active", comment="状态: active/acknowledged/resolved/ignored")
    acknowledged_by = Column(Integer, ForeignKey("users.id"), comment="确认人")
    acknowledged_at = Column(DateTime, comment="确认时间")
    ack_remark = Column(Text, comment="确认备注")
    resolved_by = Column(Integer, ForeignKey("users.id"), comment="解决人")
    resolved_at = Column(DateTime, comment="解决时间")
    resolve_remark = Column(Text, comment="解决备注")
    resolve_type = Column(String(20), comment="解决类型: manual/auto/timeout")
    duration_seconds = Column(Integer, comment="持续时间(秒)")
    is_notified = Column(Boolean, default=False, comment="是否已通知")
    notify_count = Column(Integer, default=0, comment="通知次数")
    created_at = Column(DateTime, default=datetime.now, comment="创建时间")


class AlarmRule(Base):
    """告警规则表（复合告警）"""
    __tablename__ = "alarm_rules"

    id = Column(Integer, primary_key=True, autoincrement=True)
    rule_name = Column(String(100), nullable=False, comment="规则名称")
    rule_type = Column(String(20), comment="规则类型: and/or/sequence")
    condition_expr = Column(Text, comment="条件表达式")
    alarm_level = Column(String(20), comment="告警级别")
    alarm_message = Column(String(200), comment="告警消息")
    is_enabled = Column(Boolean, default=True, comment="是否启用")
    created_at = Column(DateTime, default=datetime.now, comment="创建时间")


class AlarmShield(Base):
    """告警屏蔽表"""
    __tablename__ = "alarm_shields"

    id = Column(Integer, primary_key=True, autoincrement=True)
    point_id = Column(Integer, ForeignKey("points.id"), comment="点位ID(空表示全局)")
    alarm_level = Column(String(20), comment="屏蔽级别(空表示全部)")
    start_time = Column(DateTime, nullable=False, comment="开始时间")
    end_time = Column(DateTime, nullable=False, comment="结束时间")
    reason = Column(Text, comment="屏蔽原因")
    created_by = Column(Integer, ForeignKey("users.id"), comment="创建人")
    created_at = Column(DateTime, default=datetime.now, comment="创建时间")


class AlarmDailyStats(Base):
    """告警统计表（按天聚合）"""
    __tablename__ = "alarm_daily_stats"

    id = Column(Integer, primary_key=True, autoincrement=True)
    stat_date = Column(Date, nullable=False, comment="统计日期")
    point_id = Column(Integer, comment="点位ID")
    alarm_level = Column(String(20), comment="告警级别")
    total_count = Column(Integer, default=0, comment="总数")
    ack_count = Column(Integer, default=0, comment="已确认数")
    resolve_count = Column(Integer, default=0, comment="已解决数")
    avg_duration_seconds = Column(Integer, comment="平均持续时间")
    max_duration_seconds = Column(Integer, comment="最大持续时间")
    created_at = Column(DateTime, default=datetime.now, comment="创建时间")

File: frontend/src/api/modules/alarm.ts
/**
 * 告警管理 API
 */
import request from '@/utils/request'
import type { PageParams, PageResponse, TimeRangeParams, ExportParams } from './types'

export interface AlarmInfo {
  id: number
  alarm_no: string
  point_id: number
  point_code: string
  point_name: string
  threshold_id: number
  alarm_level: 'critical' | 'major' | 'minor' | 'info'
  alarm_type: 'threshold' | 'communication' | 'system'
  alarm_message: string
  trigger_value: number
  threshold_value: number
  status: 'active' | 'acknowledged' | 'resolved' | 'ignored'
  acknowledged_by: number | null
  acknowledged_at: string | null
  ack_remark: string | null
  resolved_by: number | null
  resolved_at: string | null
  resolve_remark: string | null
  resolve_type: 'manual' | 'auto' | 'timeout' | null
  duration_seconds: number | null
  is_notified: boolean
  notify_count: number
  created_at: string
  escalation_count?: number
  escalated_from?: string | null
  escalation_remark?: string | null
  data_source?: string
}

export interface AlarmCount {
  critical: number
  major: number
  minor: number
  info: number
  total: number
}

export interface AlarmStatistics {
  total: number
  by_level: AlarmCount
  by_type: Record<string, number>
  by_status: Record<string, number>
  by_point: { point_id: number; point_name: string; count: number }[]
  avg_duration: number
  mttr: number // Mean Time To Repair
}

export interface AlarmTrend {
  date: string
  critical: number
  major: number
  minor: number
  info: number
  total: number
}

export interface AlarmAcknowledgeParams {
  remark?: string
}

export interface AlarmResolveParams {
  resolve_type?: 'manual' | 'auto' | 'timeout'
  remark?: string
}

export interface AlarmProcessParams {
  process_remark: string
}

/**
 * 获取告警列表
 */
export function getAlarmList(params?: PageParams & TimeRangeParams & {
  alarm_level?: string
  alarm_type?: string
  status?: string
  point_id?: number
  device_type?: string
  keyword?: string
}): Promise<PageResponse<AlarmInfo>> {
  return request.get('/v1/alarms', { params })
}

/**
 * 获取活动告警
 */
export function getActiveAlarms(params?: {
  alarm_level?: string
  point_id?: number
}): Promise<AlarmInfo[]> {
  return request.get('/v1/alarms/active', { params })
}

/**
 * 获取告警详情
 */
export function getAlarmById(id: number): Promise<AlarmInfo> {
  return request.get(`/v1/alarms/${id}`)
}

/**
 * 确认告警
 */
export function acknowledgeAlarm(id: number, data?: AlarmAcknowledgeParams): Promise<void> {
  return request.put(`/v1/alarms/${id}/acknowledge`, data)
}

/**
 * 解决告警
 */
export function resolveAlarm(id: number, data?: AlarmResolveParams): Promise<void> {
  return request.put(`/v1/alarms/${id}/resolve`, data)
}

/**
 * 处理告警
 */
export function processAlarm(id: number, data: AlarmProcessParams): Promise<void> {
  return request.put(`/v1/alarms/${id}/process`, data)
}

/**
 * 批量确认告警
 */
export function batchAcknowledgeAlarms(ids: number[], remark?: string): Promise<void> {
  return request.put('/v1/alarms/batch-acknowledge', { alarm_ids: ids, remark })
}

/**
 * 获取各级别告警数量
 */
export function getAlarmCount(): Promise<AlarmCount> {
  return request.get('/v1/alarms/count')
}

/**
 * 获取告警统计
 */
export function getAlarmStatistics(params?: TimeRangeParams): Promise<AlarmStatistics> {
  return request.get('/v1/alarms/statistics', { params })
}

/**
 * 获取告警趋势
 */
export function getAlarmTrend(params?: TimeRangeParams & {
  granularity?: 'hour' | 'day' | 'week' | 'month'
}): Promise<AlarmTrend[]> {
  return request.get('/v1/alarms/trend', { params })
}

/**
 * 获取高频告警点位
 */
export function getTopAlarmPoints(params?: TimeRangeParams & {
  limit?: number
}): Promise<{ point_id: number; point_name: string; count: number }[]> {
  return request.get('/v1/alarms/top-points', { params })
}

/**
 * 导出告警记录
 */
export function exportAlarms(params?: ExportParams & {
  alarm_level?: string
  alarm_type?: string
  status?: string
}): Promise<Blob> {
  return request.get('/v1/alarms/export', {
    params,
    responseType: 'blob'
  })
}

// ==================== 告警规则管理 ====================

export interface AlarmRuleInfo {
  id: number
  rule_name: string
  rule_type: 'and' | 'or' | 'sequence'
  condition_expr: string | null
  alarm_level: 'critical' | 'major' | 'minor' | 'info'
  alarm_message: string | null
  is_enabled: boolean
  created_at: string
}

export interface AlarmRuleCreateParams {
  rule_name: string
  rule_type?: 'and' | 'or' | 'sequence'
  condition_expr?: string
  alarm_level?: 'critical' | 'major' | 'minor' | 'info'
  alarm_message?: string
  is_enabled?: boolean
}

export interface AlarmRuleUpdateParams {
  rule_name?: string
  rule_type?: 'and' | 'or' | 'sequence'
  condition_expr?: string
  alarm_level?: 'critical' | 'major' | 'minor' | 'info'
  alarm_message?: string
  is_enabled?: boolean
}

/**
 * 获取告警规则列表
 */
export function getAlarmRules(params?: PageParams & {
  rule_type?: string
  alarm_level?: string
  is_enabled?: boolean
}): Promise<PageResponse<AlarmRuleInfo>> {
  return request.get('/v1/alarms/rules', { params })
}

/**
 * 创建告警规则
 */
export function createAlarmRule(data: AlarmRuleCreateParams): Promise<AlarmRuleInfo> {
  return request.post('/v1/alarms/rules', data)
}

/**
 * 获取告警规则详情
 */
export function getAlarmRuleById(id: number): Promise<AlarmRuleInfo> {
  return request.get(`/v1/alarms/rules/${id}`)
}

/**
 * 更新告警规则
 */
export function updateAlarmRule(id: number, data: AlarmRuleUpdateParams): Promise<AlarmRuleInfo> {
  return request.put(`/v1/alarms/rules/${id}`, data)
}

/**
 * 删除告警规则
 */
export function deleteAlarmRule(id: number): Promise<void> {
  return request.delete(`/v1/alarms/rules/${id}`)
}

/**
 * 切换告警规则启用状态
 */
export function toggleAlarmRule(id: number): Promise<{ message: string; is_enabled: boolean }> {
  return request.put(`/v1/alarms/rules/${id}/toggle`)
}

// ==================== 告警屏蔽管理 ====================

export interface AlarmShieldInfo {
  id: number
  point_id: number | null
  point_code: string | null
  point_name: string | null
  alarm_level: 'critical' | 'major' | 'minor' | 'info' | null
  start_time: string
  end_time: string
  reason: string | null
  created_by: number | null
  creator_name: string | null
  created_at: string
  status: 'active' | 'expired' | null
}

export interface AlarmShieldCreateParams {
  point_id?: number | null
  alarm_level?: 'critical' | 'major' | 'minor' | 'info' | null
  start_time: string
  end_time: string
  reason?: string
}

/**
 * 获取告警屏蔽列表
 */
export function getAlarmShields(params?: PageParams & {
  point_id?: number
  alarm_level?: string
}): Promise<PageResponse<AlarmShieldInfo>> {
  return request.get('/v1/alarms/shields', { params })
}

/**
 * 创建告警屏蔽
 */
export function createAlarmShield(data: AlarmShieldCreateParams): Promise<AlarmShieldInfo> {
  return request.post('/v1/alarms/shields', data)
}

/**
 * 删除告警屏蔽
 */
export function deleteAlarmShield(id: number): Promise<void> {
  return request.delete(`/v1/alarms/shields/${id}`)
}

// ==================== 告警升级规则管理 ====================

export interface AlarmEscalationInfo {
  id: number
  rule_name: string
  source_level: string
  timeout_minutes: number
  target_level: string
  notify_user_ids: number[]
  is_enabled: boolean
  description: string | null
  escalation_chain: string | null
  created_at: string | null
  updated_at: string | null
}

export interface AlarmEscalationCreateParams {
  rule_name: string
  source_level: string
  timeout_minutes: number
  target_level: string
  notify_user_ids: number[]
  is_enabled?: boolean
  description?: string
  escalation_chain?: string
}

export interface AlarmEscalationUpdateParams {
  rule_name?: string
  source_level?: string
  timeout_minutes?: number
  target_level?: string
  notify_user_ids?: number[]
  is_enabled?: boolean
  description?: string
  escalation_chain?: string
}

export function getEscalations(params?: { source_level?: string; is_enabled?: boolean; page?: number; page_size?: number }) {
  return request.get('/v1/escalations', { params })
}

export function createEscalation(data: AlarmEscalationCreateParams) {
  return request.post('/v1/escalations', data)
}

export function getEscalation(id: number) {
  return request.get(`/v1/escalations/${id}`)
}

export function updateEscalation(id: number, data: AlarmEscalationUpdateParams) {
  return request.put(`/v1/escalations/${id}`, data)
}

export function deleteEscalation(id: number) {
  return request.delete(`/v1/escalations/${id}`)
}

export function toggleEscalation(id: number) {
  return request.put(`/v1/escalations/${id}/toggle`)
}
